doc/design/doi-generation-1.md

```diff
@@ -0,0 +1,26 @@
+# DOI generation process
+
+## dandi-api
+
+The DOI generation process happens inside a call to `POST /dandisets/{dandiset_id}/draft/publish/`.
+DOI generation is the first step of the publish process, because it is the only step to rely on an external service.
+If the call to the Datacite API fails, there are no files in S3 that need to be cleaned up.
+
+The API will pass the following to the DOI generator:
+* A DOI identifier. This will be of the form `f'{prefix}/{dandiset_id}/{version_id}'`, for example `10.80507/000001/0.210402.1835`.
+* A URL for the dandiset. This will be a redirector URL of the form `f'https://dandiarchive.org/dandiset/{dandiset_id}/{version_id}'`.
+* The dandiset metadata, as a python dictionary.
+
+## dandi.core.doi ?
+
+This function will generate the body of the DOI and return it.
+
+TODO: fill me in
```

Comment:
* checking `meta = DandisetMeta(meta)` or `meta = PublishedDandisetMeta (meta)` (if `publishedBy` and `datePublished` is available):
      - if fields are missing (e.g. `contributors`) it would return error even today
      - I would add checks if `doi`  (and possibly other fields) is available
*  running `datacite_json = to_datacite(meta)`
      - this should not return an error

Reply:
On the previous design doc we were using the GitHub suggestion feature to inject changes rather than writing PRs for the PR:
```suggestion
* checking meta = DandisetMeta(meta) or meta = PublishedDandisetMeta (meta) (if publishedBy and datePublished is available):
  * if fields are missing (e.g. contributors) it would return error even today
  * I would add additional checks if doi (and possibly other fields) is available
* running datacite_json = to_datacite(meta)
  * this should not return an error
```